src/bt.rs

```diff
@@ -0,0 +1,358 @@
+//! tree.insert(bits!(), '@');
+//! tree.insert(bits!(0), 'a');
+//! tree.insert(bits!(1), 'b');
+//! tree.insert(bits!(0, 0), 'c');
+//! tree.insert(bits!(0, 1), 'd');
+//! tree.insert(bits!(1, 0), 'e');
+//! tree.insert(bits!(1, 1), 'f');
+//! ```
+
+use core::{
+    fmt::{Debug, Display, Formatter},
+    mem,
+};
+use std::io::Cursor;
+
+use bitvec::slice::BitSlice;
+
+use crate::codec::{CodecError, Decode, Encode};
+
+/// Errors triggered by binary tree operations.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum BinaryTreeError<V> {
+    /// Error when an operation cannot reach the end of a path.
+    #[error("unreachable path in the tree")]
+    UnreachablePath(V),
+}
+
+/// Used to indicate a traversal path on the binary tree.
+pub type Path = BitSlice;
+
+type SubTree<V> = Option<Box<Node<V>>>;
+
+/// Represents a node of a binary tree.
+struct Node<V> {
+    value: V,
+    left: SubTree<V>,
+    right: SubTree<V>,
+}
+
+impl<V> Node<V> {
+    fn new(value: V) -> Self {
+        Self {
+            value,
+            left: None,
+            right: None,
+        }
+    }
+}
+/// Represents an append-only binary tree.
+pub struct BinaryTree<V> {
+    root: SubTree<V>,
+}
+
+impl<V> BinaryTree<V> {
+    /// Creates an empty binary tree.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Inserts the value at the end of the path, possibly replacing an existing value.
+    ///
+    /// # Returns
+    /// - `Ok(None)`, the node was inserted at the end of the path.
+    /// - `Ok(Some(old))`, the `old` value at the end of the path was replaced.
+    /// - `Err(UnreachablePath(value))`, the end of the path is unreachable.
+    ///    No insertion occurs and the value is returned to its owner.
+    pub fn insert(&mut self, path: &Path, value: V) -> Result<Option<V>, BinaryTreeError<V>> {
+        let mut node = &mut self.root;
+
+        // Finds the node at the end of the path.
+        for bit in path {
+            if let Some(next_node) = node {
+                node = if !bit {
+                    &mut next_node.left
+                } else {
+                    &mut next_node.right
+                };
+            } else {
+                return Err(BinaryTreeError::UnreachablePath(value));
+            }
+        }
+
+        // Checks whether the node already has a value,
+        // If so, it replaces with the new value, and returns the old one.
+        // otherwise, node is available for insertion, and `None` is returned.
+        if let Some(current) = node {
+            Ok(Some(mem::replace(&mut current.value, value)))
+        } else {
+            *node = Some(Box::new(Node::new(value)));
+            Ok(None)
+        }
+    }
+
+    /// Gets a reference to the value located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
```

Comment:
It might be a good idea to return a reference to the node rather than just the value. That way we can start an insertion from this point if we wanted to. I think we'll end up wanting to do this in VIDPF.

Reply:
added a new method.